File: src/leetcode/python/greedy/JumpGame_55.py
class JumpGame_55(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        length = len(nums)
        if length == 1:
            return True

        max_pos = -1
        for i, n in enumerate(nums):
            max_pos = max(i + n, max_pos)
            if n == 0 and i >= max_pos:
                break
            if max_pos >= length - 1:
                return True
        return False

File: src/leetcode/java/JumpGame_55.java
package leetcode.java.medium;

public class JumpGame_55 {
  public boolean canJump(int[] nums) {
    if (nums.length == 1) return true;

    int maxPos = 0;
    for (int i = 0; i < nums.length; i++) {
      if (nums[i] == 0 && i > maxPos) break;
      maxPos = Math.max(i + nums[i], maxPos);
      if (maxPos >= nums.length - 1) return true;
    }
    return false;
  }
}
